feat(lifecycle): document waiting for sidecar to be ready (#1449)
kumahq/kuma-website: app/_src/production/dp-config/dpp-on-kubernetes.md

To join the mesh in a graceful way, we need to first make sure the application is ready to serve traffic before it can be considered a valid traffic destination.

> {% if_version lte:2.3.x %}

Reviewer: ```suggestion
{% if_version gte:2.4.x %}
```

> If any of the init containers experience network connectivity issues we suggest trying setting `runtime.kubernetes.injector.sidecarContainer.waitForDataplaneReady` to `true` so that the init container for the sidecar finishes only when the sidecar is ready to serve traffic.

Reviewer: There's also an annotation that can be added on a per pod basis. We should make sure it's documented here and in   the annotations page.

Reviewer: Maybe it's worth explaining that by default all containers start in parallel and you have no guarantee that the application won't start sending requests before the sidecar is ready. This would make these requests fail.

If this is required you can use the configuration ...

WDYT? Explain the problem before the fix

Author: > all containers start in parallel

they don't - https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#differences-from-regular-containers

> If you specify multiple init containers for a Pod, kubelet runs each init container sequentially. 

they run sequentially, that's why the thing that Lukasz implemented works.

Reviewer: The doc you linked to are for init-containers right?
it might be that they don't start in parallel but they don't wait for the previous one to be ready before starting the next one...

{% warning %}

> The `waitForDataplaneReady` setting relies on Kubernetes container creation sequence (which is undocumented, so it might be changed in the future) and injecting kuma-init as the first init container (which we can't guarantee, because other mutating webhooks can rearrange the init containers) with a `postStart` hook.

Reviewer: I'd rewrite this warning like this:

```
The waitForDataplaneReady setting relies on the way that Kubernetes creates containers, which is not documented and could change in the future. It also relies on the Kuma init container being injected as the first init container in the pod, which is not guaranteed because other mutating webhooks can rearrange the init containers.

A better solution will be available when sidecar containers are more stable and widely available.
```

Parentheses are usually better to be avoided.

When `Pod` is converted to a `Dataplane` object it will be marked as unhealthy until Kubernetes considers all containers to be ready.

{% if_version gte:2.4.x %}
### Waiting for the dataplane to be ready

> By default, containers are started in any order, so an application container can start even though a dataplane container might not be ready to receive traffic.

Author: ```suggestion
By default, containers start in any order, so an app container can start even though a dataplane container might not be ready to receive traffic.
```

> Making initial requests, such as connecting to a database, can fail for a brief period after the pod is started. ```

Author: ```suggestion
Making initial requests, such as connecting to a database, can fail for a brief period after the pod starts. 
```

> This can make initial requests (like connecting to a database) fail for short time after starting the pod.

Author: ```suggestion
Making initial requests, such as connecting to a database, can fail for a brief period after the pod is started. ```

To mitigate this problem we suggest setting
* `runtime.kubernetes.injector.sidecarContainer.waitForDataplaneReady` to `true`, or 
* [kuma.io/wait-for-dataplane-ready](/docs/{{ page.version }}/reference/kubernetes-annotations/#wait-for-dataplane-ready) annotation to `true`
so that the app container waits for the dataplane container to be ready to serve traffic.

{% warning %}

> The `waitForDataplaneReady` setting relies on the fact that when a `postStart` hook is defined Kubernetes runs containers in sequence of occurrence in `containers` list.

Author: ```suggestion
The `waitForDataplaneReady` setting relies on the fact that defining a `postStart` hook causes Kubernetes to run containers sequentially based on their order of occurrence in the `containers` list.
```

This is not documented and could change in the future.
It also relies on the kuma-sidecar container being injected as the first container in the pod, which is not guaranteed because other mutating webhooks can rearrange the containers.

> A better solution will be available when [sidecar containers](https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/) are more stable and widely available.

Author: ```suggestion
As [sidecar containers](https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/) gain stability and wider availability, a better solution is going to be implemented.
```

> A better solution will be available when [sidecar containers](https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/) are more stable and widely available.

Author: ```suggestion
<!-- vale off -->
A better solution will be available when [sidecar containers](https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/) are more stable and widely available.
<!-- vale on -->
```

> As [sidecar containers](https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/) gain stability and wider availability, a better solution is going to be implemented.

Author: ```suggestion
A better solution will be available when [sidecar containers](https://kubernetes.io/blog/2023/08/25/native-sidecar-containers/) are more stable and widely available.
```
